### PR TITLE
Maintain topology service

### PR DESCRIPTION
**Pull request by samt1803**
Added maintainTopology service, helper and tests to operatorplugin

### CODE REVIEW

File: packages/broker/src/plugins/operator/MaintainTopologyHelper.ts [hunk 1]
```diff
@@ -17,23 +17,23 @@ export interface OperatorClientEvents {
...
 
     /**
      * Emitted when a unstaked from ALL Sponsorships for the given stream
      */
-    removeStakedStream: (streamId: string, blockNumber: number) => void
+    removeStakedStream: (streamId: string) => void
 }
 
-export interface OperatorClientConfig {
+export interface TopologyHelperConfig {
```
Comment (harbu) [inline packages/broker/src/plugins/operator/MaintainTopologyHelper.ts]: For consistency I would name:
```suggestion
export interface MaintainTopologyHelperConfig {
```

File: packages/broker/src/plugins/operator/MaintainTopologyHelper.ts [hunk 2]
```diff
@@ -6,7 +6,7 @@ import { EventEmitter } from "eventemitter3"
 import { FetchResponse, Logger, TheGraphClient } from "@streamr/utils"
 
 /**
- * Events emitted by {@link OperatorClient}.
+ * Events emitted by {@link MaintainTopologyHelper}.
  */
 export interface OperatorClientEvents {
```
Comment (harbu) [inline packages/broker/src/plugins/operator/MaintainTopologyHelper.ts]: For consistency I would name:
```suggestion
export interface MaintainTopologyHelperEvents {
```

File: packages/broker/src/plugins/operator/MaintainTopologyHelper.ts [hunk 3]
```diff
@@ -93,22 +93,27 @@ export class OperatorClient extends EventEmitter<OperatorClientEvents> {
...
             if (sponsorshipCount === 0) {
                 this.sponsorshipCountOfStream.delete(streamId)
-                this.emit("removeStakedStream", streamId, await this.contract.provider.getBlockNumber())
+                this.emit("removeStakedStream", streamId)
             }
+            latestBlock = await this.contract.provider.getBlockNumber()
         })
-        await this.pullStakedStreams()
+        
+        const initalStreams = await this.pullStakedStreams(latestBlock)
```
Comment (harbu) [inline packages/broker/src/plugins/operator/MaintainTopologyHelper.ts:90]: I believe `latestBlock` would always equal `0` at this point due to any potential event handlers getting fired earliest after the synchronous portion of the `start` method has finished executing. So a potential bug here?

```suggestion
        const initalStreams = await this.pullStakedStreams(latestBlock)
```

File: packages/broker/src/plugins/operator/MaintainTopologyHelper.ts [hunk 4]
```diff
@@ -93,22 +93,27 @@ export class OperatorClient extends EventEmitter<OperatorClientEvents> {
...
 
-    async getStakedStreams(): Promise<string[]> {
-        return Array.from(this.sponsorshipCountOfStream.keys())
-    }
+    // async getStakedStreams(): Promise<string[]> {
+    //     return Array.from(this.sponsorshipCountOfStream.keys())
+    // }
 
-    private async pullStakedStreams(): Promise<{ streamIds: string[], blockNumber: number }> {
+    private async pullStakedStreams(requiredBlocknumber: number): Promise<string[]> {
```
Comment (harbu) [inline packages/broker/src/plugins/operator/MaintainTopologyHelper.ts]:
```suggestion
    private async pullStakedStreams(requiredBlockNumber: number): Promise<string[]> {
```

File: packages/broker/src/plugins/operator/MaintainTopologyHelper.ts [hunk 5]
```diff
@@ -93,22 +93,27 @@ export class OperatorClient extends EventEmitter<OperatorClientEvents> {
...
         const bounty = new Contract(sponsorshipAddress, sponsorshipABI, this.contract.provider as Provider) as unknown as Sponsorship
         return bounty.streamId()
     }
 
-    async getStakedStreams(): Promise<string[]> {
-        return Array.from(this.sponsorshipCountOfStream.keys())
-    }
+    // async getStakedStreams(): Promise<string[]> {
+    //     return Array.from(this.sponsorshipCountOfStream.keys())
+    // }
```
Comment (harbu) [inline packages/broker/src/plugins/operator/MaintainTopologyHelper.ts]: Remove?

File: packages/broker/src/plugins/operator/MaintainTopologyHelper.ts [hunk 6]
```diff
@@ -133,16 +138,17 @@ export class OperatorClient extends EventEmitter<OperatorClientEvents> {
                     `
             }
         }
-        let latestBlockNumber = 0
+        // let latestBlockNumber = 0
```
Comment (harbu) [inline packages/broker/src/plugins/operator/MaintainTopologyHelper.ts]: Remove commented out code

File: packages/broker/src/plugins/operator/MaintainTopologyService.ts [hunk 7]
```diff
@@ -1,10 +1,10 @@
 import { Logger, Multimap } from '@streamr/utils'
-import { OperatorClient } from './OperatorClient'
+import { MaintainTopologyHelper } from './MaintainTopologyHelper'
 import StreamrClient, { Stream, Subscription } from 'streamr-client'
 import { StreamID, StreamPartIDUtils, toStreamID } from '@streamr/protocol'
 import { SetMembershipSynchronizer } from '../storage/SetMembershipSynchronizer'
 import pLimit from 'p-limit'
-import { compact } from 'lodash'
+// import { compact } from 'lodash'
```
Comment (harbu) [inline packages/broker/src/plugins/operator/MaintainTopologyService.ts]:
```suggestion
```

File: packages/broker/src/plugins/operator/MaintainTopologyService.ts [hunk 8]
```diff
@@ -16,72 +16,51 @@ function toStreamIDSafe(input: string): StreamID | undefined {
     }
 }
 
-function singletonSet<T>(element: T): Set<T> {
-    return new Set<T>([element])
-}
+// function singletonSet<T>(element: T): Set<T> {
+//     return new Set<T>([element])
+// }
```
Comment (harbu) [inline packages/broker/src/plugins/operator/MaintainTopologyService.ts]:
```suggestion
```

File: packages/broker/src/plugins/operator/MaintainTopologyService.ts [hunk 9]
```diff
@@ -16,72 +16,51 @@ function toStreamIDSafe(input: string): StreamID | undefined {
...
 export class MaintainTopologyService {
     private readonly streamrClient: StreamrClient
-    private readonly operatorClient: OperatorClient
+    private readonly maintainTopologyHelper: MaintainTopologyHelper
     private readonly subscriptions = new Multimap<StreamID, Subscription>()
     private readonly synchronizer = new SetMembershipSynchronizer<StreamID>()
     private readonly concurrencyLimit = pLimit(1)
 
-    constructor(streamrClient: StreamrClient, operatorClient: OperatorClient) {
+    constructor(streamrClient: StreamrClient, operatorClient: MaintainTopologyHelper) {
```
Comment (harbu) [inline packages/broker/src/plugins/operator/MaintainTopologyService.ts]:
```suggestion
    constructor(streamrClient: StreamrClient, maintainTopologyHelper: MaintainTopologyHelper) {
```

File: packages/broker/src/plugins/operator/MaintainTopologyService.ts [hunk 10]
```diff
@@ -16,72 +16,51 @@ function toStreamIDSafe(input: string): StreamID | undefined {
...
         this.streamrClient = streamrClient
-        this.operatorClient = operatorClient
+        this.maintainTopologyHelper = operatorClient
```
Comment (harbu) [inline packages/broker/src/plugins/operator/MaintainTopologyService.ts]:
```suggestion
        this.maintainTopologyHelper = maintainTopologyHelper
```

File: packages/broker/test/unit/plugins/operator/MaintainTopologyService.test.ts [hunk 11]
```diff
@@ -94,49 +94,32 @@ describe('MaintainTopologyService', () => {
     it('ignores non-existing streams on start', async () => {
         await setUpAndStart([STREAM_A, STREAM_NOT_EXIST, STREAM_C])
 
-        expect(streamrClient.subscribe).toHaveBeenCalledTimes(1 + 2)
+        // expect(streamrClient.subscribe).toHaveBeenCalledTimes(1 + 2)
```
Comment (harbu) [inline packages/broker/test/unit/plugins/operator/MaintainTopologyService.test.ts:97]:
```suggestion
```

File: packages/broker/test/integration/plugins/operator/smartContractUtils.ts [hunk 12]
```diff
@@ -11,19 +11,32 @@ import { Chains } from '@streamr/config'
...
-export async function generateWalletWithGasAndTokens(provider: JsonRpcProvider): Promise<Wallet> {
+export async function generateWalletWithGasAndTokens(provider: Provider): Promise<Wallet> {
     const newWallet = new Wallet(fastPrivateKey())
     const adminWallet = new Wallet(ADMIN_WALLET_PK).connect(provider)
     const token = getTokenContract().connect(adminWallet)
-    await (await token.transfer(newWallet.address, parseEther("1000"))).wait()
+    for (let i = 0; i < 5; i++) {
+        try {
+            // eslint-disable-next-line no-console
+            // console.log("trying with nonce " + await adminWallet.getTransactionCount() + " time " + new Date().getTime() / 1000)
```
Comment (harbu) [inline packages/broker/test/integration/plugins/operator/smartContractUtils.ts:29]:
```suggestion
```

File: packages/broker/test/integration/plugins/operator/smartContractUtils.ts [hunk 13]
```diff
@@ -11,19 +11,32 @@ import { Chains } from '@streamr/config'
...
+            await (await token.transfer(newWallet.address, parseEther("1000"), {
+                nonce: await adminWallet.getTransactionCount()
+            })).wait()
+            break
+        } catch (e) {
+            await new Promise((resolve) => setTimeout(resolve, 3000))
+            // eslint-disable-next-line no-console
+            // console.log("sending link failed, retrying")
```
Comment (harbu) [inline packages/broker/test/integration/plugins/operator/smartContractUtils.ts:37]:
```suggestion
```

File: packages/broker/test/integration/plugins/operator/MaintainTopologyService.test.ts [hunk 14]
```diff
@@ -77,5 +77,6 @@ describe('MaintainTopologyService', () => {
         await (await operatorContract.unstake(sponsorship1.address)).wait()
         await waitForCondition(async () => (await client.getSubscriptions()).length === 3)
         expect(await getSubscribedStreamPartIds(client)).toEqual(stream2.getStreamParts())
+        service.stop()
```
Comment (harbu) [inline packages/broker/test/integration/plugins/operator/MaintainTopologyService.test.ts]: I believe it gets stopped in `afterEach` on line 41 already?

File: packages/broker/test/integration/plugins/operator/MaintainTopologyHelper.test.ts [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+import { generateWalletWithGasAndTokens } from "./smartContractUtils"
+
+const config = Chains.load()["dev1"]
+const theGraphUrl = `http://${process.env.STREAMR_DOCKER_DEV_HOST ?? '10.200.10.1'}:8000/subgraphs/name/streamr-dev/network-subgraphs`
+
+const logger = new Logger(module)
+
+jest.setTimeout(60 * 1000)
+
+describe("MaintainTopologyHelper", () => {
```
Comment (harbu) [inline packages/broker/test/integration/plugins/operator/MaintainTopologyHelper.test.ts:28]: Can we remove commented out code here?

File: packages/broker/src/plugins/operator/MaintainTopologyService.ts [hunk 16]
```diff
@@ -16,72 +16,51 @@ function toStreamIDSafe(input: string): StreamID | undefined {
...
-            return
-        }
-        if (this.subscriptions.get(streamId).length > 0) {
-            logger.warn('Ignore already subscribed stream', { streamId, blockNumber })
-            return
-        }
-        await this.addStream(streamId, blockNumber)
-    })
+    private onAddStakedStreams = async (streamIDs: string[]) => {
+        streamIDs.map(this.parseStreamIdWrapper(this.addStream.bind(this)))
```
Comment (harbu) [inline packages/broker/src/plugins/operator/MaintainTopologyService.ts:43]: Method `parseStreamIdWrapper` probably needs rework to simplify this, but I can look at that in a follow-up PR.

File: packages/broker/src/plugins/operator/MaintainTopologyHelper.ts [hunk 17]
```diff
@@ -76,10 +68,10 @@ export class OperatorClient extends EventEmitter<OperatorClientEvents> {
             const sponsorshipCount = (this.sponsorshipCountOfStream.get(streamId) || 0) + 1
             this.sponsorshipCountOfStream.set(streamId, sponsorshipCount)
             if (sponsorshipCount === 1) {
-                this.emit("addStakedStream", streamId, await this.contract.provider.getBlockNumber())
+                this.emit("addStakedStream", [streamId])
             }
+            latestBlock = await this.contract.provider.getBlockNumber()
```
Comment (harbu) [inline packages/broker/src/plugins/operator/MaintainTopologyHelper.ts]: I think this line is not needed. We would've have already initiated execution of method `pullStakedStreams` when this update would occur, therefore not affecting anything.

File: packages/broker/src/plugins/operator/MaintainTopologyHelper.ts [hunk 18]
```diff
@@ -93,22 +85,23 @@ export class OperatorClient extends EventEmitter<OperatorClientEvents> {
             this.sponsorshipCountOfStream.set(streamId, sponsorshipCount)
             if (sponsorshipCount === 0) {
                 this.sponsorshipCountOfStream.delete(streamId)
-                this.emit("removeStakedStream", streamId, await this.contract.provider.getBlockNumber())
+                this.emit("removeStakedStream", streamId)
             }
+            latestBlock = await this.contract.provider.getBlockNumber()
```
Comment (harbu) [inline packages/broker/src/plugins/operator/MaintainTopologyHelper.ts]: Same comment as above

> I think this line is not needed. We would've have already initiated execution of method `pullStakedStreams` when this update would occur, therefore not affecting anything.

File: packages/broker/src/plugins/operator/MaintainTopologyService.ts [hunk 19]
```diff
@@ -16,72 +13,49 @@ function toStreamIDSafe(input: string): StreamID | undefined {
...
-}
-
 export class MaintainTopologyService {
     private readonly streamrClient: StreamrClient
-    private readonly operatorClient: OperatorClient
+    private readonly maintainTopologyHelper: MaintainTopologyHelper
     private readonly subscriptions = new Multimap<StreamID, Subscription>()
     private readonly synchronizer = new SetMembershipSynchronizer<StreamID>()
     private readonly concurrencyLimit = pLimit(1)
+    private readonly logger: Logger
```
Comment (harbu) [inline packages/broker/src/plugins/operator/MaintainTopologyService.ts]: I would keep the static logger pattern as before. So
```ts
const logger = new Logger(module)
```
at the beginning of this file and then just use that. That way the log message have a clear module name in their statements, e.g.
```
INFO [2023-06-19T07:20:06.010] (MaintainTopologyService  ): Logging something
```

File: packages/broker/src/plugins/operator/MaintainTopologyService.ts [hunk 20]
```diff
@@ -16,72 +13,49 @@ function toStreamIDSafe(input: string): StreamID | undefined {
...
-        await this.operatorClient.start()
-        const rawStreamIds = await this.operatorClient.getStakedStreams()
-        const blockNumber = 10 // TODO: fix properly
-        const streamIds = new Set(compact([...rawStreamIds].map(toStreamIDSafe)))
-        const { added } = this.synchronizer.ingestSnapshot(streamIds, blockNumber)
-        for (const streamId of added) {
-            await this.addStream(streamId, blockNumber)
-        }
-        logger.info('Started')
+        this.logger.info('Starting MaintainTopologyService')
```
Comment (harbu) [inline packages/broker/src/plugins/operator/MaintainTopologyService.ts]:
```suggestion
        this.logger.info('Starting')
```

That way the log message (using the static logger) would become:
```
INFO [2023-06-19T07:20:06.010] (MaintainTopologyService  ): Starting
```
(No need to re-state the class name)

File: packages/broker/src/plugins/operator/MaintainTopologyService.ts [hunk 21]
```diff
@@ -16,72 +13,49 @@ function toStreamIDSafe(input: string): StreamID | undefined {
...
+        this.maintainTopologyHelper.on('addStakedStream', this.onAddStakedStreams)
+        this.maintainTopologyHelper.on('removeStakedStream', this.onRemoveStakedStream)
+        await this.maintainTopologyHelper.start()
+        this.logger.info('Started MaintainTopologyService')
```
Comment (harbu) [inline packages/broker/src/plugins/operator/MaintainTopologyService.ts]:
```suggestion
        this.logger.info('Started')
```

File: packages/broker/src/plugins/operator/MaintainTopologyService.ts [hunk 22]
```diff
@@ -16,72 +13,49 @@ function toStreamIDSafe(input: string): StreamID | undefined {
...
     }
 
-    // eslint-disable-next-line class-methods-use-this
     async stop(): Promise<void> {
-        this.operatorClient.stop()
-        logger.info('stopped')
+        this.maintainTopologyHelper.stop()
+        this.logger.info('stopped')
```
Comment (harbu) [inline packages/broker/src/plugins/operator/MaintainTopologyService.ts]:
```suggestion
        this.logger.info('Stopped')
```
Capitalize in accordance with https://github.com/streamr-dev/network/blob/main/internal-docs/logging-practices.md

File: packages/broker/src/plugins/operator/MaintainTopologyHelper.ts (hunk 17 above)
Comment (samt1803) [reply packages/broker/src/plugins/operator/MaintainTopologyHelper.ts]: indeed!

**Review by harbu — APPROVED** [review]
lgtm!

**Pull request merged**